[Security] Add Guard authenticator <supports> method

Amo opened the pull request:
This method will be called before starting an authentication against a guard authenticator.
The authentication will be tried only if the supports method returned `true`
This improves understanding of code, increase consistency and removes responsability for `getCredentials` method to decide if the current request should be supported or not.

| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | yes |
| Deprecations? | yes |
| Tests pass? | yes |
| Fixed tickets | none |
| License | MIT |
| Doc PR |  |

Amo committed:
[Security] Add Guard authenticator <supports> method

This method will be called before starting an authentication against a guard authhenticator.
The authentication will be tried only if the supports method returned <true>
This improves understanding of code, increase consistency and removes responsability for <getCredentials> method
To decide if the current request should be supported or not.

Amo commented:
Note: as being not familiar with the recent deprecation mechanisms, i'm not sure i did it well in `src/Symfony/Component/Security/Guard/Firewall/GuardAuthenticationListener.php`, line 109

Amo commented:
About tests, on php 5.5.30 i get 

```
$ ./phpunit --filter GuardAuthenticationListenerTest
PHPUnit 4.8.19 by Sebastian Bergmann and contributors.

Testing Symfony Test Suite
.......

Time: 4.1 seconds, Memory: 173.00Mb

OK (7 tests, 22 assertions)

Remaining deprecation notices (6)

Class Mock_GuardAuthenticatorInterface_a5f0cf1d should provide a method <supports(Request $request)>. see GuardAuthenticatorSupportInterface: 6x
    2x in GuardAuthenticationListenerTest::testReturnNullToSkipAuth from Symfony\Component\Security\Guard\Tests\Firewall
    1x in GuardAuthenticationListenerTest::testHandleCatchesAuthenticationException from Symfony\Component\Security\Guard\Tests\Firewall
    1x in GuardAuthenticationListenerTest::testHandleSuccessStopsAfterResponseIsSet from Symfony\Component\Security\Guard\Tests\Firewall
    1x in GuardAuthenticationListenerTest::testHandleSuccess from Symfony\Component\Security\Guard\Tests\Firewall
    1x in GuardAuthenticationListenerTest::testHandleSuccessWithRememberMe from Symfony\Component\Security\Guard\Tests\Firewall

KO symfony
```

otherwise, tests passes on 5.6 & 7.0

iltar commented on src/Symfony/Component/Security/Guard/Firewall/GuardAuthenticationListener.php:
```diff
@@ -94,12 +94,37 @@ private function executeGuardAuthenticator($uniqueGuardKey, GuardAuthenticatorIn
+            if (method_exists($guardAuthenticator, 'supports')) {
+                if (true !== $guardAuthenticator->supports($request)) {
+                    return;
+                }
```
You add an interface but you check if the method exists here. Why not simply check if the interface is in use?

Amo replied:
That was because i did not  handled it this way originally. In fact now it would make more sense to check the interface implementation.

Amo commented:
by the way @weaverryan , that's what i was talking with you this morning.

stof commented:
I'm not sure we need this. It would probably have to duplicate some of the work done in getCredentials in many time.

@weaverryan what do you think ?

stof commented:
@Amo tests are failing because you are using the deprecated way in a non-legacy test.

weaverryan commented:
It's "clearer" in some ways because supports() can return a clean true/false value and then getCredentials() can cause an AuthException if it returns null, which would make it more consistent with getUser and checkCredentials.

I discussed this at the conference - I'm mixed on it - the only negative I see is exactly what Stof mentions, but that is a practical concern. For an API token auth, I'll either have to duplicate code that gets/checks for the API token value in supports() and getUser() or just always return true from supports, which defeats the purpose.

@amo Can you convince us that this will be a better user experience? 

Thx!

Amo commented:
The idea is to separate the responsabilities of 
* Identifying if the authentication context is elligible/supported
* Extracting the credentials from the context

and doing so, improve the readability, support of variations and testability of an authentication method.

`Supports` would be responsible to detect a specific context, guaranteeing that when `getCredentials` is called, this context has been met.

As @stof said, it could end-up in duplicating the same checks in `supports` and `getCredentials`. But, imho, that would occurs if the authentication system supports many context and is poorly designed.

Let's take an example :

Access to `/api` requires a `X-API-Token` header.
An authenticator is created :

```PHP
    // ...

    /**
     * Ensure the given request provides an API token in the headers
     * 
     * @param  Request $request
     * 
     * @return null|mixed
     */
    public function supports(Request $request)
    {
        if (false === $request->headers->has('X-API-Token')) {
            return false;
        }

        return true;
    }

    /**
     * Extract the credentials from the request.
     * 
     * @param  Request $request
     * 
     * @return string
     */
    public function getCredentials(Request $request)
    {
        return $request->headers->get('X-API-Token'));
    }

```

If additional credentials retrieval methods are required (qs, payload ...) it is likely that the `getCredentials` method will get many `if` and `else`, which would make the code harder to read, maintain, test ... 

This separation `supports` vs `getCredentials` will spur the developer to create dedicated guard authenticator for each specificities. A common abstract class or a trait could be used to share the common stages of the authentication. But concrete final classes will explicitly indicates with use-cases they manage : 

```
Security
└── Authenticator
    ├── ApiBaseAuthenticator.php
    ├── ApiHeaderAuthenticator.php
    ├── ApiPayloadAuthenticator.php
    └── ApiQueryAuthenticator.php
```

Where `ApiBaseAuthenticator.php` handles the common logic for 

```
getUser
checkCredentials
createAuthenticatedToken
onAuthenticationFailure
onAuthenticationSuccess
supportsRememberMe
```

And each dedicated authenticator provides the dedicated logic to retrieve credentials from header, query string or payload

Amo commented:
> @Amo tests are failing because you are using the deprecated way in a non-legacy test.

How can i avoid it ? googling about Symfony 3 and deprecation conventions didn't give any results.

weaverryan commented:
@Amo I think your arguments are quite good. So, let's see if we can complete this PR :).

Forgetting about BC, how would you want the interface to look in 4.0? You added a new `GuardAuthenticatorSupportsInterface`, but I imagine that you'd really want `supports()` to be in the GuardAuthenticatorInterface, correct? There will be some work that will need to be done to create a new interface (that has the new method) and deprecate the old interface.

weaverryan commented:
@Amo about the tests, the GuardAuthenticationListenerTest is mocking the GuardAuthenticatorInterface, which then triggers your deprecation warning. This mock should be changed to `GuardAuthenticatorSupportsInterface` so that the deprecated functionality isn't triggered (or changed to whatever final interface we decide on that has the new `supports()` method). There should also be a test that verifies that if you use the original `GuardAuthenticatorInterface`, then the old behavior is used. Put an `@legacy` above tests that specifically test how legacy functionality works.

splio-aleroux commented:
ok thanks, i'm on it

Amo commented:
ok thanks, i'm on it

Amo commented:
Tests are still running.
But, (take a deep breath), i moved the interface into the Authenticator sub namespace, `Symfony\Component\Security\Guard\Authenticator\GuardAuthenticatorInterface` which makes sense to me. 
This interface adds the `supports` method and keep the same the signature as `Symfony\Component\Security\Guard\GuardAuthenticatorInterface`. It extends `Symfony\Component\Security\Guard\GuardAuthenticatorInterface`as `GuardAuthenticatorLegacyInterface` to avoid BC break. the `Symfony\Component\Security\Guard\GuardAuthenticatorInterface` has now a `@deprecated` annotation and is still used as type hint in `Symfony\Component\Security\Guard\Firewall\GuardAuthenticationListener` alongside the new `Symfony\Component\Security\Guard\Authenticator\GuardAuthenticatorInterface` interface.

Amo commented:
Should i fixup and `git push --force` all the changes on the Amo:symphony branch ?

iltar commented:
I notice the `handle()`/`executeGuardAuthenticator()` are becoming big*(ger with those changes)*. I think the flow and changes are good, but this will become harder and harder to maintain. Some suggestions I have for this method which could be done (*after* this PR is merged):

**Using a null logger instead of a PSR logger**
It will improve readability and reduce complexity and would remove 8 if statements in just those 2 methods

**Move code out of the try catch which is unrelated**
This try catch is pretty much just a sort of "failsafe" at the moment. However, things like `supports()` should not throw an `AuthenticationException`.  A list of method calls:
 - `$this->logger->debug('...');`
 - `$guardAuthenticator->supports($request)`
 - `$guardAuthenticator->getCredentials($request)`
 - `$this->logger->debug('...');`
 - `$this->authenticationManager->authenticate($token)`
 - `$this->logger->info('...');`
 - `$this->guardHandler->authenticateWithToken($token, $request);`

Out of all available methods in the authenticator, only `getUser()` and `checkCredentials()` throw an exception and they are not being called here. In fact, the only method that throws an exception is `$this->authenticationManager->authenticate($token)`. It's actually making it possible for people to throw all kinds of `AuthenticationExceptions` which can cause unwanted behavior or break when someone decides this try/catch can be narrowed down in say 3.4.

**Make `GuardAuthenticationListener` final**
We've created a very nice extension point which people _will_ abuse in the future. My suggestion is to make this class final as it makes it easier for people to fix things here. Behavioral changes should still be avoided as much as possible though. 


After this the code will still be quite massive but it will at least be easier to read and understand what it does.

fabpot commented:
@weaverryan What's the status of this one?

fabpot commented:
@weaverryan I'd like to move this one forward. Can you have a look at it? Thanks.

stof commented on src/Symfony/Component/Security/Guard/Firewall/GuardAuthenticationListener.php:
```diff
@@ -94,12 +95,32 @@ private function executeGuardAuthenticator($uniqueGuardKey, GuardAuthenticatorIn
+                }
+
+                // otherwise something went wrong and the authentication must fail
+                throw new \InvalidArgumentException('Guard authenticator must provide user credentials');
```
A RuntimeException (or even better its child UnexpectedValueException) would be better. This is not an invalid argument here, but an invalid return value

weaverryan replied:
And with the message:

>
Did you forget to return some credentials from Foo\Bar\ClassName::getCredentials()? Credentials can be anything: an array, string or object.

weaverryan commented on src/Symfony/Component/Security/Guard/Authenticator/GuardAuthenticatorInterface.php:
```diff
@@ -0,0 +1,168 @@
+     *
+     * @return bool
+     */
+    public function supports(Request $request);
```
Move this to the top - the order is significant for clarity

weaverryan replied:
What about? `shouldAttemptAuthentication(Request $request)`?

Amo replied:
`shouldAttemptAuthentication ` is less clear to me. I prefer simple method name in a narrow scope.

weaverryan commented on src/Symfony/Component/Security/Guard/Authenticator/GuardAuthenticatorInterface.php:
```diff
@@ -0,0 +1,168 @@
+     *
+     * If it does, authentication based on the given request will continue.
+     * If not the next (if any) authenticator will be requested
+     * until at one supports the current request.
```
>
If this returns true, authentication will continue (e.g. getCredentials() will be called).
If false, this authenticator is done. The next (if any) authenticators will be called and
may authenticate the user, or leave the user as anonymous.

weaverryan commented on src/Symfony/Component/Security/Guard/Firewall/GuardAuthenticationListener.php:
```diff
@@ -11,16 +11,17 @@
+use Symfony\Component\Security\Guard\GuardAuthenticatorHandler;
+use Symfony\Component\Security\Guard\Authenticator\GuardAuthenticatorInterface;
+use Symfony\Component\Security\Guard\GuardAuthenticatorInterface as GuardAuthenticatorLegacyInterface;
+use Symfony\Component\Security\Guard\Token\PreAuthenticationGuardToken;
```
Unless there's a standard I don't know of, can you move the `use` statements back to their original spot? It's nit-picky I know, but it will help avoid conflicts in the future.

Amo replied:
My SDK did re-order them in alphabetical order.
This has a benefit in term of readability but i can revert this change indeed.

Amo commented:
Updated.

weaverryan commented on src/Symfony/Component/Security/Guard/Authenticator/GuardAuthenticatorInterface.php:
```diff
@@ -0,0 +1,168 @@
+     *
+     * @return mixed|null
+     */
+    public function getCredentials(Request $request);
```
@Amo What do you think about actually removing `getCredentials()` now? And instead changing `getUser()` to be `getUser(Request $request, UserProviderInterface $userProvider);`

Now that getCredentials() *only* has the responsibility of returning the credentials, it seems simpler just to ask "give me a user" instead of breaking things into these 2 pieces.

Btw, if you agree, we will need to rename getUser() to something like `fetchUser()` so we can add it to the new interface.

Amo replied:
It could be simpler, but it would add much more complexity in the `fetchUser` method.
This "merge" of the two responsabilities will make the `fetchUser` method a bit more complex to unit test.
Aditionnaly, the `UserProviderInterface` is known by the `GuardAuthenticationProvider` and here we are in the `GuardAuthenticationListener` so it needs many changes.

weaverryan replied:
It's just a question of "DX", which is subjective:

Before:
```php
public function getCredentials(Request $request)
{
    return $request->headers->get('Authorization');
}

public function getUser($credentials, UserProviderInterface $userProvider)
{
    $authorizationHeader = $credentials;
    // ...
}
```

After

```php
public function fetchUser(Request $request, UserProviderInterface $userProvider)
{
    $authorizationHeader = $request->headers->get('Authorization');
    // ...
}
```

@javiereguiluz @iltar Do you guys have a preference?

@Amo This new method would be called from GuardAuthenticationProvider - we would set the entire request in the `PreAuthenticationGuardToken` so that we could pass it to the new method.

iltar replied:
I'm not 100% sure about the "fetch" part, it deviates from the currently popular find and get (and makes me feel like talking to my dog). DX wise I'd go for the second solution, however the first option gives you the ability to make a chain credential provider in symfony. That might be overkill and people can still implement a similar pattern in their `fetchUser`. If `getCredentials` only purpose is to be passed to `getUser`, they should be merged imo.

weaverryan replied:
Then how about we name it `findUser()`? I agree with you on the chaining - there *may* be a use for this, but I was also thinking: implement it yourself.

iltar replied:
Personally I'd lean more towards `getUser()`, this because it's rather confusing with the Repositories otherwise. By default those start with `find` and it might confuse developers when those names start getting mixed up. Ideally I'd throw an exception in a `get*` method if unable to get whatever was expected. 

How would that work out if the could would always throw an exception if the user could not be found?

weaverryan replied:
@iltar The problem with `getUser()` is that we already have a method named that - and this would cause a signature change. So we need a new method name so we can keep the old one for BC. So, it's not idea. Perhaps `getUserFromRequest()`?

iltar replied:
It's starting to sound more like a builder pattern;

```php
$guard->buildUser(new UserFromRequestBuilder($request), $userProvider);

// potential to decouple from the request
$guard->buildUser(new UserFromConsoleBuilder($input, $output), $userProvider);
```

Anyway, `getUserFromRequest()` sounds decent enough, what will happen to `getUser()`?

fabpot replied:
We can re-use the same method name. With `get_func_args()`, it's possible to detect which version the user is using, like we've done in the past elsewhere in the framework. I'd go with the current `getUser()` method name.

weaverryan commented on src/Symfony/Component/Security/Guard/Firewall/GuardAuthenticationListener.php:
```diff
@@ -94,12 +96,35 @@ private function executeGuardAuthenticator($uniqueGuardKey, GuardAuthenticatorIn
+                // authenticator is expected to give not-null credentials.
+                $credentialsCanBeNull = false;
+            } else {
+                // it's a GuardAuthenticatorLegacyInterface
```
We should trigger the deprecation notice here. For example: https://github.com/symfony/symfony/blob/2.8/src/Symfony/Component/Security/Core/SecurityContext.php#L14

Amo replied:
Triggered line 111

weaverryan commented:
Thanks for the quick update @Amo - and sorry for the extra comments - I had to pause in the middle of reviewing :). Btw, when I talked about Guard at the last conference, someone else suggested this exact change afterwards.

Amo commented:
I've been updating the code according to the discussion about `getUserFromRequest`.
Unit tests need to be updated.

weaverryan commented:
Thanks @Amo!

So, we have two choices with this:

1) Revert your last commit (my suggestions related to `getUserFromRequest`) and finish the original intentions of this PR.

2) Keep going with `getUserFromRequest`

My concern is that it will take some work to get `getUserFromRequest` to be BC with all of the proper deprecation warnings. If you're willing to iterate on that with me over the next 2 weeks, let's do it! Or, we can go a slightly easier route and keep `getUserFromRequest` until later.

Let me know - and I'll help push this along either way. Thanks for your work so far!

splio-aleroux commented:
My main issue here is that I don't have much time in the coming days/week. 
Reverting the last commit and iterate later could be a safer option

Amo commented:
Edit : wrong user. 

My main issue here is that I don't have much time in the coming days/week. 
Reverting the last commit and iterate later could be a safer option

iltar commented:
@weaverryan what would have to be done for `getUserFromRequest()`?

weaverryan commented:
Thanks @Amo (about the wrong user).

@iltar it's "easy" in theory - even to get it working *both* ways (the old and new way). The tricky part is getting the deprecation notices in the write spot and making sure the new interface is correct. Basically, not breaking BC and triggering deprecation is the tough part (but less tough for someone experienced in doing that).

weaverryan commented:
@Amo let's do as you said then, revert that last commit and get this done for the original idea. If we can do it quickly enough, I may be able to iterate *again* on the next solution... but this gives us at least the first step.

Amo commented:
Last commit removed.

weaverryan commented on src/Symfony/Component/Security/Guard/Authenticator/GuardAuthenticatorInterface.php:
```diff
@@ -0,0 +1,168 @@
+interface GuardAuthenticatorInterface extends GuardAuthenticatorLegacyInterface
+{
+    /**
+     * Does the authenticator supports the given Request ?
```
`support`

weaverryan commented on src/Symfony/Component/Security/Guard/Firewall/GuardAuthenticationListener.php:
```diff
@@ -94,12 +96,36 @@ private function executeGuardAuthenticator($uniqueGuardKey, GuardAuthenticatorIn
+
+                // otherwise something went wrong and the authentication must fail
+                throw new \UnexpectedValueException(sprintf(
+                    'Did you forget to return some credentials %s::getCredentials()',
```
> You must return some credentials from %s:getCredentials(). To
> skip authentication, return false from %s::supports().

weaverryan commented:
@Amo I just left some minor comments - but it's just meaningless wording stuff. Update that when you get a chance, but...

:+1: ping @symfony/deciders The behavior looks good to me, but I'd like a double-check that the deprecation of the interface was done correctly.

Status: Reviewed

splio-aleroux commented:
Wording changed.

Amo commented:
Wording updated.

xabbuh commented:
Some tests need to be updated as they don't handle the deprecations properly.

Status: Needs work

weaverryan commented on src/Symfony/Component/Security/Guard/Tests/Firewall/GuardAuthenticationListenerTest.php:
```diff
@@ -184,6 +204,45 @@ public function testHandleCatchesAuthenticationException()
+
+        $authenticatorA
+            ->expects($this->never())
+            ->method('supports');
```
This test fails on php7 with:

> Symfony\Component\Security\Guard\Tests\Firewall\GuardAuthenticationListenerTest::testLegacyInterfaceAvoidSupportsCall
PHPUnit_Framework_MockObject_RuntimeException:
> Trying to configure method "supports" which cannot be configured because it does not exist,
> has not been specified, is final, or is static

I'm not sure you can mock a method that doesn't exist. But, I'm not sure this is necessary at all: if `supports()` *were* accidentally called on this mock, wouldn't it explode with a "Method supports() does not exist" anyways?

xabbuh replied:
Yeah, we need to change that. The recent version of PHPUnit's mock package now warns when you try to mock invalid methods and we therefore fixed some tests this week. Can you point me to the code this test should cover?

Amo replied:
It tests [this condition](https://github.com/Amo/symfony/blob/master/src/Symfony/Component/Security/Guard/Firewall/GuardAuthenticationListener.php#L100) is not executed when using the legacy interface

xabbuh replied:
Ah I see, the solution is to create a class that implements the legacy interface which changes some value in its `supports()` method and then check that the value was never changed (thus meaning that `supports()` wasn't executed).

xabbuh replied:
see https://github.com/symfony/symfony/pull/18284/files#diff-d8949c4b7c75325b0bef282e1d2041b9R1235 for an example

splio-aleroux replied:
I get the idea. I'll do this asap.

weaverryan commented on src/Symfony/Component/Security/Guard/Tests/Firewall/GuardAuthenticationListenerTest.php:
```diff
@@ -184,6 +204,45 @@ public function testHandleCatchesAuthenticationException()
+    }
+
+    /**
+     * @legacy
```
Should be `@group legacy` - sorry I missed that at first!

weaverryan commented on src/Symfony/Component/Security/Guard/Tests/Firewall/GuardAuthenticationListenerTest.php:
```diff
@@ -184,6 +204,45 @@ public function testHandleCatchesAuthenticationException()
     }
 
+    /**
+     * @legacy
```
same here

weaverryan commented:
@amo I just left some very minor comments, which will likely fix the failing tests :)

Amo commented:
I added the `@group legacy` annotations instead of `@legacy`
I changed the unit tests to avoid testing that the method `supports` is not called on legacy interface as it doesn't exists.

Removed test : `testLegacyInterfaceAvoidSupportsCall`
Added tests : 
  * `testLegacyInterfaceNullCredentials`
  * `testLegacyInterfaceKeepsWorking`

trousers commented on src/Symfony/Component/Security/Guard/Firewall/GuardAuthenticationListener.php:
```diff
@@ -86,7 +88,7 @@ public function handle(GetResponseEvent $event)
     }
 
-    private function executeGuardAuthenticator($uniqueGuardKey, GuardAuthenticatorInterface $guardAuthenticator, GetResponseEvent $event)
+    private function executeGuardAuthenticator($uniqueGuardKey, GuardAuthenticatorLegacyInterface $guardAuthenticator, GetResponseEvent $event)
```
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [The method executeGuardAuthenticator() has a Cyclomatic Complexity of 14. The configured cyclomatic complexity threshold is 10.](https://www.codacy.com/app/git_9/symfony/file/2331610611/issues/source?bid=3207821&fileBranchId=3207918#l91)

trousers replied:
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [The method executeGuardAuthenticator() has an NPath complexity of 304. The configured NPath complexity threshold is 200.](https://www.codacy.com/app/git_9/symfony/file/2331610611/issues/source?bid=3207821&fileBranchId=3207918#l91)

iltar replied:
@trousers can you please stop it? You're spamming my mailbox. If you want to propose this tool, please open an issue instead.

weaverryan commented:
The one remaining test failure looks totally unrelated (from the Process component). 

Status: Reviewed

xabbuh commented on src/Symfony/Component/Security/Guard/Firewall/GuardAuthenticationListener.php:
```diff
@@ -94,12 +96,38 @@ private function executeGuardAuthenticator($uniqueGuardKey, GuardAuthenticatorIn
+                // it's a GuardAuthenticatorLegacyInterface
+                // we support the previous behaviour to avoid BC break.
+                $credentialsCanBeNull = true;
+                @trigger_error('The Symfony\Component\Security\Guard\GuardAuthenticatorInterface interface is deprecated since version 3.0 and will be removed in 4.0. Use Symfony\Component\Security\Guard\Authenticator\GuardAuthenticatorInterface instead.', E_USER_DEPRECATED);
```
3.1 instead of 3.0

HeahDude replied:
Could be
```php
@trigger_error(sprintf'The "%s" interface is deprecated since version 3.1 and will be removed in 4.0. Use "%s" instead.', GuardAuthenticatorLegacyInterface::class, GuardAuthenticatorInterface::class), E_USER_DEPRECATED);

HeahDude commented on src/Symfony/Component/Security/Guard/Authenticator/GuardAuthenticatorInterface.php:
```diff
@@ -0,0 +1,168 @@
+     *
+     * @param mixed         $credentials
+     * @param UserInterface $user
+     *
```
`@return bool` ?

Amo replied:
I just added the `@return bool` but after reading with more attention the description, i would suggest to replace it by `@return mixed`. Only `true` has a meaning, anything else, that can be returned, will make the authentication failed. So this method does not return only boolean. By the way, this is copied from the legacy interface. @weaverryan

HeahDude commented on src/Symfony/Component/Security/Guard/Authenticator/GuardAuthenticatorInterface.php:
```diff
@@ -0,0 +1,168 @@
+    public function checkCredentials($credentials, UserInterface $user);
+
+    /**
+     * Create an authenticated token for the given user.
```
typo "Create**s**"

HeahDude commented on src/Symfony/Component/Security/Guard/Authenticator/GuardAuthenticatorInterface.php:
```diff
@@ -0,0 +1,168 @@
+     *  C) The "remember me" functionality is activated. This is usually
+     *      done by having a _remember_me checkbox in your form, but
+     *      can be configured by the "always_remember_me" and "remember_me_parameter"
+     *      parameters under the "remember_me" firewall key
```
Is this the expected indentation ?

Amo replied:
This is copied from the LegacyInterface. @weaverryan

HeahDude commented on src/Symfony/Component/Security/Guard/Firewall/GuardAuthenticationListener.php:
```diff
@@ -94,12 +96,38 @@ private function executeGuardAuthenticator($uniqueGuardKey, GuardAuthenticatorIn
                 $this->logger->debug('Calling getCredentials() on guard configurator.', array('firewall_key' => $this->providerKey, 'authenticator' => get_class($guardAuthenticator)));
             }
 
+            // abort the execution of the authenticator if it doesn't supports the request.
```
typo "support"

HeahDude commented:
Nice changes :) Thanks!

Amo commented:
Is there anything else expected here ?

fabpot commented:
This looks good, but I don't like the fact that we introduce a brand new interface with the same name but in a sub-namespace, that looks like a hack to me.

romain-pierre commented:
Why not allow the `supports` method to determine if the user is authenticated? In many use cases, the authenticators are called even if the user is already authenticated.

chalasr commented:
Same concern as @fabpot, adding a sub interface with only `supports()` is probably the only alternative (can't help for naming, sorry).

Amo commented:
Sure, i'll do my best

Amo commented:
I've been _finally_ able to work on this !

`Symfony\Component\Security\Guard\Authenticator\GuardAuthenticatorInterface` has been moved/renamed to `Symfony\Component\Security\Guard\AuthenticatorInterface`

weaverryan commented on src/Symfony/Component/Security/Guard/GuardAuthenticatorInterface.php:
```diff
@@ -28,6 +28,8 @@
  *
  * @author Ryan Weaver <ryan@knpuniversity.com>
+ *
+ * @deprecated Symfony\Component\Security\Guard\AuthenticatorInterface must be used instead
```
Do we need to `trigger_error` inside this file as well (and if so, we would need to copy this interface into `AuthenticatorInterface` so that using the new interface doesn't trigger the warning). I'm not sure because, we WILL trigger the deprecation notice the minute you try to use this interface in the guard system... so in practice... it's covered.

chalasr replied:
AFAIK we don't, the Debug component triggers automatically based on the `@deprecated` annotation

weaverryan reviewed (approved):
This looks good to me! 👍 I left one question about a possible deprecation that should be added, but I'm not 100% that it should be added.

weaverryan commented:
Just a note: we may have missed feature freeze for 3.4 on this. But, I *do* think it's ready. If we *were* to delay this to 4.1, we could/should merge early in the process. @chalasr and I are also interested in investigating another change to the interface (removing the user provider). And so, doing that all at once in 4.1 has its advantages.

fabpot commented:
As a French guy, I don't think we missed anything. In the past, we merged some "old" PRs a week after code freeze. I think it's fair to give a bit more time for older PRs.

sroze commented:
That would be arsh to refuse 3.4's milestone for a few days to a PR created in 2015 :smiley:

chalasr commented:
@sroze The issue is that any change made to an interface is a BC break since interfaces cannot trigger deprecations. Here, we have no other choice that deprecating the whole existing interface to replace it by a new one, just for introducing one method. And we won't do that twice, forcing users to switch to a new interface on the same code for two consecutive majors sucks, there's not enough good names also.

@Amo We asked you to finish this for 3.4 and you did it. But, because we have another change in mind for the same interface (the one Ryan is talking about), I think it would be good to postpone this for 4.1 so that the interface switch happens only once. Apologies for that, it's all about improving guard authenticators API. This is 👍  for me as is, we just need a bit more time.

weaverryan commented:
Btw, the issue we're talking about is #21391. 

But I just talked with @chalasr about this. We're +1 for merging to 3.4. Sorry for the back and forth.

There *is* another change we may want to make to the guard interface, but it deals with user providers  making them optional. It's a very complex topic, and I'm afraid we will "tie up" this PR with a big, complex task that may not get done (I hope it does, but let's be pragmatic).

So, a +1 for me

chalasr reviewed (approved):
👍  for 3.4, we're not ready for the user provider topic.

chalasr commented:
@Amo Could you add a note in UPGRADE-3.4, UPGRADE-4.0 and Security CHANGELOG files? I'll do it tomorrow otherwise

chalasr reviewed (changes requested):
Ah, I see a few `GuardAuthenticatorInterface` references that should be updated (https://github.com/symfony/symfony/search?utf8=%E2%9C%93&q=GuardAuthenticatorInterface). Also includes `AbstractGuardAuthenticator` which should stop implementing the deprecated interface, and probably provide a `supports() { return true; }` method. I'll also look at this tomorrow if you can't do it before.

Amo commented:
I won't be able to perform these changes tonight.
I'll do my best to work on it tomorrow morning.

Amo commented:
@chalasr @weaverryan I've applied the last changes requested.
It implies that the inheritance between `AuthenticatorInterface` and `GuardAuthenticatorInterface` is inverted and by default `AbstractGuardAuthenticator` provides a concrete implementation of the `supports` method.

Amo commented:
In addition i've added a deprecation in `AbstractGuardAuthenticator::supports` method as it should not be used except for backward compatibility.

Amo commented:
The error on AppVeyor is weirdly unrelated (Form component)

Amo commented:
I've updated the UPGRADE files. CHANGELOG files do not exist yet for 3.4 / 4.0 unless i'm wrong.

chalasr commented on src/Symfony/Component/Security/Guard/Provider/GuardAuthenticationProvider.php:
```diff
@@ -142,6 +142,6 @@ private function authenticateViaGuard(GuardAuthenticatorInterface $guardAuthenti
     public function supports(TokenInterface $token)
     {
-        return $token instanceof GuardTokenInterface;
+        return $token instanceof TokenInterface;
```
I think this should be reverted

nicolas-grekas replied:
why this change?

chalasr commented on src/Symfony/Component/Security/Guard/AbstractGuardAuthenticator.php:
```diff
@@ -19,9 +20,25 @@
+     * @param Request $request
+     *
+     * @return bool
+     */
```
Should be
```php
/**
 * {@inheritdoc}
 * 
 * @deprecated since version 3.4, to be removed in 4.0
 */
```

chalasr commented on UPGRADE-4.0.md:
```diff
@@ -643,6 +643,9 @@ Security
    class instead.
 
+ * The `GuardAuthenticatorInterface` has been removed. Implement
+   `Symfony\Component\Security\Guard\AuthenticatorInterface` class instead.
```
`class` should be removed

chalasr commented on src/Symfony/Component/Security/Guard/AbstractGuardAuthenticator.php:
```diff
@@ -19,9 +20,25 @@
+     */
+    public function supports(Request $request)
+    {
+        @trigger_error('The Symfony\Component\Security\Guard\AbstractGuardAuthenticator::supports default implementation is used. This is provided for backward compatibility on GuardAuthenticationInterface that is deprecated since version 3.1 and will be removed in 4.0. Provide your own implementation of the supports method instead.', E_USER_DEPRECATED);
```
```php
@trigger_error(sprintf('The "%s()" method is deprecated since version 3.4 and will be removed in 4.0. Implement the "%s::supports()" method in class "%s" instead.', __METHOD__, AuthenticatorInterface::class, get_class($this)));
```

chalasr commented on src/Symfony/Component/Security/Guard/Firewall/GuardAuthenticationListener.php:
```diff
@@ -92,20 +94,46 @@ public function handle(GetResponseEvent $event)
+                // it's a GuardAuthenticatorInterface
+                // we support the previous behaviour to avoid BC break.
+                $credentialsCanBeNull = true;
+                @trigger_error('The Symfony\Component\Security\Guard\GuardAuthenticatorInterface interface is deprecated since version 3.1 and will be removed in 4.0. Use Symfony\Component\Security\Guard\Authenticator\GuardAuthenticatorInterface instead.', E_USER_DEPRECATED);
```
This should be removed as it triggers an extra deprecation (just checked your PR, tests remain green after removing it, avoiding 4 unexpected deprecations notices).

chalasr commented on src/Symfony/Component/Security/Guard/Firewall/GuardAuthenticationListener.php:
```diff
@@ -92,20 +94,46 @@ public function handle(GetResponseEvent $event)
+            // abort the execution of the authenticator if it doesn't support the request.
+            if ($guardAuthenticator instanceof GuardAuthenticatorInterface) {
+                // it's a GuardAuthenticatorInterface
+                // we support the previous behaviour to avoid BC break.
```
Please replace this comment by `deprecated since version 3.4, to be removed in 4.0`, we basically `grep`on that sentence for deprecating remove deprecated features

chalasr reviewed (changes requested):
Last changes look good to me. Some last comments below, we're almost good.

chalasr commented:
@amo I fixed the remaining comments on my side but I can't push to your fork. 
Could you [enable push permissions for maintainers on your fork](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/)) or hard reset your branch on https://github.com/chalasr/symfony/tree/guard-supports ?
Other alternative is to close this PR and reopen a new one from my fork, but it would be too bad to lose the discussion of this PR.

Amo commented:
Hi, thank you for your changes. 
I've enable commits from maintainers on this PR

chalasr reviewed (approved):
👍

nicolas-grekas commented on UPGRADE-3.4.md:
```diff
@@ -294,7 +294,10 @@ Security
  * Deprecated the HTTP digest authentication: `NonceExpiredException`,
    `DigestAuthenticationListener` and `DigestAuthenticationEntryPoint` will be
-   removed in 4.0. Use another authentication system like `http_basic` instead.
+   removed in 4.0. Use another authentication system like `http_basic` instead.   
```
extract space at end of line (same on the line below)

nicolas-grekas commented on src/Symfony/Component/Security/Guard/Firewall/GuardAuthenticationListener.php:
```diff
@@ -92,20 +94,45 @@ public function handle(GetResponseEvent $event)
+            // abort the execution of the authenticator if it doesn't support the request.
+            if ($guardAuthenticator instanceof GuardAuthenticatorInterface) {
+                // deprecated since version 3.4, to be removed in 4.0
+                @trigger_error(sprintf('The %s interface is deprecated since version 3.4 and will be removed in 4.0. Implement %s instead.', GuardAuthenticatorInterface::class, AuthenticatorInterface::class), E_USER_DEPRECATED);
```
this runtime deprecation should be removed: it has already been trigger by DebugClassLoader

nicolas-grekas commented on src/Symfony/Component/Security/Guard/Firewall/GuardAuthenticationListener.php:
```diff
@@ -92,20 +94,45 @@ public function handle(GetResponseEvent $event)
+                }
+
+                // otherwise something went wrong and the authentication must fail
+                throw new \UnexpectedValueException(sprintf(
```
should be on one line

nicolas-grekas commented on src/Symfony/Component/Security/Guard/GuardAuthenticatorHandler.php:
```diff
@@ -61,14 +61,14 @@ public function authenticateWithToken(TokenInterface $token, Request $request)
      * @return null|Response
      */
-    public function handleAuthenticationSuccess(TokenInterface $token, Request $request, GuardAuthenticatorInterface $guardAuthenticator, $providerKey)
+    public function handleAuthenticationSuccess(TokenInterface $token, Request $request, AuthenticatorInterface $guardAuthenticator, $providerKey)
```
that's a BC break unfortunately (same for all public interfaces)

nicolas-grekas commented on src/Symfony/Component/Security/Guard/GuardAuthenticatorInterface.php:
```diff
@@ -28,8 +29,10 @@
+ * @deprecated since version 3.4, to be removed in 4.0. Use AuthenticatorInterface instead.
  */
-interface GuardAuthenticatorInterface extends AuthenticationEntryPointInterface
+interface GuardAuthenticatorInterface extends AuthenticatorInterface
```
I think this should be done the other way around, to still allow passing legacy type-hints.
this would mean removing the runtime notice above, but DebugClassLoader will take care or triggering it anyway

nicolas-grekas replied:
`AuthenticatorInterface extends GuardAuthenticatorInterface`

chalasr commented on src/Symfony/Component/Security/Guard/Provider/GuardAuthenticationProvider.php:
```diff
@@ -56,7 +56,7 @@ public function __construct($guardAuthenticators, UserProviderInterface $userPro
      * Finds the correct authenticator for the token and calls it.
      *
-     * @param GuardTokenInterface $token
+     * @param TokenInterface|GuardTokenInterface $token
```
I reverted this change as it should have been `TokenInterface` from day 1. I'm going to fix it on 2.7

chalasr commented:
@nicolas-grekas comments addressed, thanks

nicolas-grekas reviewed (approved):
(but there are more FQCN strings that could be replaced by `...::class`)

chalasr committed:
Fix BC layer

chalasr commented:
Remaining FQCNs fixed

xabbuh commented on src/Symfony/Component/Security/Guard/Firewall/GuardAuthenticationListener.php:
```diff
@@ -172,12 +190,12 @@ public function setRememberMeServices(RememberMeServicesInterface $rememberMeSer
+     * @param Response               $response
      */
-    private function triggerRememberMe(GuardAuthenticatorInterface $guardAuthenticator, Request $request, TokenInterface $token, Response $response = null)
+    private function triggerRememberMe($guardAuthenticator, Request $request, TokenInterface $token, Response $response = null)
```
IMO we should be consistent in removing the type hint (like here) or not (like in `GuardAuthenticatorHandler` below). Keeping the type hint might be a good idea to catch all places where we need to adjust them when removing code in 4.0.

chalasr replied:
👍  I re-added the typehint in this class so that it breaks if we forget to change them when removing the deprecated interface.

fabpot commented:
Thank you @Amo.

Amo commented:
Thank you ! Now i'm a proud contributor of Symfony

chalasr commented:
@Amo Thanks for this and congrats for your first PR on github! You did it right, in the right repo :)

kbond commented:
I am getting what I think is a false positive deprecation notice in 3.4:

```
User Deprecated: The "Symfony\Component\Security\Guard\AbstractGuardAuthenticator::supports()" 
Method is deprecated since version 3.4, to be removed in 4.0. You should not extend it from 
"...\MyAuthenticator".
```

I have added `MyAuthenticator::supports()` but am still getting this notice. Did I not upgrade correctly or is this a bug?

kbond commented on src/Symfony/Component/Security/Guard/AbstractGuardAuthenticator.php:
```diff
@@ -19,9 +20,21 @@
     /**
+     * {@inheritdoc}
+     *
+     * @deprecated since version 3.4, to be removed in 4.0
```
If I remove this line, the false positive deprecation notice is removed. Should I open a PR?

chalasr replied:
Please do. This method must be implemented by concrete subclasses and it is unlikely to be called directly.
